google/cloud/bigquery/v2/minimal/internal/json_utils.cc

```diff
@@ -106,15 +119,11 @@ nlohmann::json RemoveJsonKeysAndEmptyFields(
               [&parsed](std::string const& key) { return parsed == key; });
           return !discard;
         }
-        if (event == nlohmann::json::parse_event_t::object_end) {
-          return parsed != nullptr && !parsed.empty();
-        }
-        if (event == nlohmann::json::parse_event_t::array_end) {
-          return !parsed.empty();
-        }
```

Comment:
I am not sure I follow the intent behind replacing this with a recursive function. Can you explain?

Reply:
Ah I see . looks like the empty objects part above may not be getting removed correctly.

The filter keys are removed in the rest_stub here : https://github.com/googleapis/google-cloud-cpp/blob/2cb3cdbf81a370601c122772db0a02c97b25daf5/google/cloud/bigquery/v2/minimal/internal/job_rest_stub.cc#L68 assuming they were set in the request.

Can you also add a check there to make sure the filtering is only called if the `json_filter_keys` in the request is= non-empty?

Reply:
It seems like changing the business logic. I'll update the behavior in the next PR with a unit test for it.

Reply:
I don't think changing like so should change the overall business logic. Filtering should only get applied if filtering keys have been set in the request.

SGTM for doing it in a follow-up PR.

```
std::string filtered_json = json_payload;
if (!request.json_filter_keys.empty()) {
filtered_json = RemoveJsonKeysAndEmptyFields(json_payload.dump(),
                                                    request.json_filter_keys());
}
```